### ISSUE TITLE
[Bug]: No signature of method is applicable for argument types

### ISSUE THREAD

**Issue by lucas404x** [fluttercommunity/plus_plugins#1722]
### Platform

Android 11 (Emulator)

### Plugin

share_plus

### Version

6.3.3

### Flutter SDK

3.10.0-1.3.pre

### Steps to reproduce

1. flutter run

### Logs

```shell
FAILURE: Build failed with an exception.

* Where:
Build file 'C:\Users\Desktop\AppData\Local\Pub\Cache\hosted\pub.dev\share_plus-6.3.3\android\build.gradle' line: 27

* What went wrong:
A problem occurred evaluating project ':share_plus'.
> No signature of method: build_982riom3gsel1z49qvhv3msqn.android() is applicable for argument types: (build_982riom3gsel1z49qvhv3msqn$_run_closure2) values: [build_982riom3gsel1z49qvhv3msqn$_run_closure2@709d9b67]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 28s
Exception: Gradle task assembleDebug failed with exit code 1
```

### Flutter Doctor

```shell
[√] Flutter (Channel beta, 3.10.0-1.3.pre, on Microsoft Windows [Version 10.0.22624.1616], locale en-US)
    • Flutter version 3.10.0-1.3.pre on channel beta at C:\Users\Desktop\fvm\versions\beta
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d11aff97d2 (7 days ago), 2023-04-19 11:11:05 -0700
    • Engine revision a645b36a9b
    • Dart version 3.0.0 (build 3.0.0-417.2.beta)
    • DevTools version 2.23.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at C:\Users\Desktop\AppData\Local\Android\sdk
    • Platform android-33, build-tools 33.0.0-rc4
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Build Tools 2019 16.11.26)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools
    • Visual Studio Build Tools 2019 version 16.11.33529.622
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2022.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[√] VS Code (version 1.77.3)
    • VS Code at C:\Users\Desktop\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.62.0

[√] Connected device (4 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64    • Microsoft Windows [Version 10.0.22624.1616]
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 112.0.5615.138
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 112.0.1722.58

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

### Checklist before submitting a bug

- [X] I Google'd a solution and I couldn't find it
- [X] I searched on StackOverflow for a solution and I couldn't find it
- [X] I read the README.md file of the plugin
- [X] I'm using the latest version of the plugin
- [X] All dependencies are up to date with `flutter pub upgrade`
- [X] I did a `flutter clean`
- [X] I tried running the example project

**Comment by nietsmmar** [comment]
I am experiencing the same issue

**Comment by tomasatisocco** [comment]
Same issue, can't build project because of this

**Comment by nohli** [comment]
Could you please share your Gradle versions? Can search for `com.android.tools.build:gradle` in `build.gradle`.

It looks like issue https://github.com/fluttercommunity/android_id/issues/38.

If your gradle versions are below 7, it could be because `namespace` was not available in prior versions. Then probably an if check would do the trick.

```
Starting with AGP 7.0, you can set the namespace in the app’s build.gradle file
```
https://developer.android.com/build/publish-library/prep-lib-release#choose-namespace

**Comment by tomasatisocco** [comment]
> Could you please share your Gradle versions? Can search for `com.android.tools.build:gradle` in `build.gradle`.
> 
> It looks like issue [fluttercommunity/android_id#38](https://github.com/fluttercommunity/android_id/issues/38).
> 
> If your gradle versions are below 7, it could be because `namespace` was not available in prior versions. Then probably an if check would do the trick.
> 
> ```
> Starting with AGP 7.0, you can set the namespace in the app’s build.gradle file
> ```
> 
> https://developer.android.com/build/publish-library/prep-lib-release#choose-namespace

classpath 'com.android.tools.build:gradle:4.1.3'

**Comment by nohli** [comment]
I've found a workaround that should work (didn't find where to check for the AGP version in `build.gradle`):

```
    // Only set namespace on compatible versions (Android Gradle plugin >= 7.0.0)
    // The if-check checks for a property that has been removed since AGP 7.0.0
    // https://developer.android.com/build/releases/past-releases/agp-7-0-0-release-notes#build-cache-removed
    if (!project.hasProperty("android.buildCacheDir")) {
        namespace = 'dev.fluttercommunity.android_id'
    }
```

**Comment by saulopef** [comment]
Same issue here, any workaround?

**Comment by vbuberen** [comment]
The problem is in `namespace` property that we added to be compatible with Android Gradle Plugin 8. This property is supported since Android Gradle Plugin 7.0 only.

You need to update your projects to use Android Gradle Plugin to be at least 7.0. I would suggest to use something like `7.4.2`. 
Here is an example on where to set the version: https://github.com/fluttercommunity/plus_plugins/blob/main/packages/package_info_plus/package_info_plus/android/build.gradle#L13

**Comment by vbuberen** [comment]
> classpath 'com.android.tools.build:gradle:4.1.3'

This is way too old. Updating the project will help not only with compatibility, but also will improve build speed for Android part of your project.

**Comment by vbuberen** [comment]
> AGP 4.2 is actually from 2020 ([there were no versions 5-6](https://developer.android.com/build/releases/gradle-plugin#versioning-update))

Thanks, I know that as still doing Android development and aware of releases.

> I think, it's not good to force users to upgrade a different dependency. Could for some users be problematic because of other dependencies as well.

It is completely fine to update Android Gradle plugin as it doesn't break anything in the project and is a simple change. It is also a good reason to modernise the project and benefit from build speed improvements which Google and Gradle teams were working on.
With addition of `namespace` we resolved the problem of compatibility with AGP 8 which is now the version forced by Google in Android Studio Flamingo released last week.

**Comment by vbuberen** [comment]
> I would suggest, either fixing it (with the changes from the branch), or making it a breaking change and add docs (for which it's too late - or we remove the latest version after the breaking change update). What do you think?

I don't want to support old AGP versions as it would encourage people to continue ignoring their Android build config.

> Actually, AGP 4.x and below are "EOL'd" for quite some time already and everyone should have migrated away by now.

This is exactly what I am talking about. It is very bad that in Flutter community developers are often not aware of such changes and just keep their projects not updated for a very long time, leading to such issues like we have now.

From one side I agree that we need to mark it as a breaking change, but as < 7 AGP is really a thing of the past I would just update our docs to mention this requirement. This case reminds me of situation we have with other plugins which easily bump min Flutter version to 3 or 3.3 without releasing a new breaking change, because, everyone understands that if project is maintained it is already using new Flutter version, not a really old one. 

Here is a few examples of such plugins without releasing a breaking change:
https://pub.dev/packages/webview_flutter/changelog#410
https://pub.dev/packages/image_picker/changelog#0872
https://pub.dev/packages/url_launcher/changelog#619

So, to sum up this discussion:
- I will update docs to mention this requirement of AGP 7.x.
- I would really love this case to make Plug plugins a push that makes people move away from AGP older than 7.x versions. It is not painful to update at all as requires to just change the version of AGP, Gradle wrapper (if it is very old as well) and (if project is really really old) update Kotlin version as versions < 1.4.x are not supported by AGP 7.x plugin. Even if it sounds like something not right it would lead to more projects being up to date and benefiting from build speeds improvements.
- We will be more сareful for future releases to avoid similar cases.

**Comment by vbuberen** [comment]
Reopening to mention it in PRs with docs updates

**Comment by vbuberen** [comment]
Will leave this here as well: https://github.com/flutter/flutter/issues/125621#issuecomment-1525500990

**Comment by stuartmorgan** [comment]
> This case reminds me of situation we have with other plugins which easily bump min Flutter version to 3 or 3.3 without releasing a new breaking change, because, everyone understands that if project is maintained it is already using new Flutter version, not a really old one.

FYI this is fundamentally different. If you update a plugin to a min Flutter version of 3.3, then someone using Flutter 3.0 just won't get that version of the plugin. When they `flutter pub get`, the resolver will only consider versions where the Flutter constraint is compatible with their version, so they will automatically get an older version that is compatible.

If you require a new version of AGP, the resolver has no idea and the plugin client will get a version that doesn't build (and for which the build error is usually *very* hard for most people to understand.)

**Comment by vbuberen** [comment]
> FYI this is fundamentally different. If you update a plugin to a min Flutter version of 3.3, then someone using Flutter 3.0 just won't get that version of the plugin. When they `flutter pub get`, the resolver will only consider versions where the Flutter constraint is compatible with their version, so they will automatically get an older version that is compatible.

Alright, didn't know it. Thanks for sharing. In such case I will do a bit differently - release a new patch version reverting the namespace addition and when we get a confirmation that workaround suggested by @nohli works will release a breaking change version of all our plugins.

**Comment by stuartmorgan** [comment]
> and when we get a confirmation that workaround suggested by @nohli works will release a breaking change version of all our plugins

A workaround that is backwards compatible shouldn't require a breaking change, since the point of backwards compatibility is that nobody is broken.

**Comment by vbuberen** [comment]
> when we get a confirmation that workaround suggested by @nohli works will release a breaking change version of all our plugins.

> A workaround that is backwards compatible shouldn't require a breaking change, since the point of backwards compatibility is that nobody is broken.

Made a typo while jumping between different projects discussions. Won't add a workaround, but make a breaking change as we would require to have AGP 7 at least.

**Comment by nohli** [comment]
How about releasing my fix for backwards-compatibility as patch-versions to fix the build for users.

And afterwards revert it and publish a breaking change?

I already released the fix for android_id and it seems fine.

**Comment by vbuberen** [comment]
No, thanks. I have already returned the previous state and don't want to use the workaround.

**Comment by vbuberen** [comment]
Updates published on pub.dev

**Comment by stuartmorgan** [comment]
FYI, we've determined that there is a version of the conditional that isn't a workaround: https://github.com/flutter/flutter/issues/125621#issuecomment-1525995461
